### ISSUE TITLE
julius -input adinnet protocol

### ISSUE THREAD

**Issue by Picolabo** [julius-speech/julius#30]
Hi,

I am trying to connect to julius server socket from a java program.

In the documentation it's written that "The protocol is a specific one sending just a sequence of audio sample streams per a small packet. There are no detailed document for the procotol, but it’s a basic and very simple one, since it has no encription or encode/decode features. adintool implements the protocol."

So I am reading adintool.c (not used to c...) and I have found adin_send_end_of_segment() that seems to send an empty char at the end of a segment, so I tried to do that. I can send a few packets (sometimes only one), but after that, I always get "Connection reset by peer: socket write error" on the client.

While I am looking into adintool.c, ....does someone out there know the detailed protocol? 

Thanks.

**Comment by Picolabo** [comment]
... and on the server, the messages I get are:

Stat: adin_tcpip: connected
Error: rdwt: transfer data length exceeded: 8977504 (>640000)
STAT: no input frame
Stat: adin_tcpip: connection end

Another thing I don't understand is that I am sure that this number of 8977504 is way bigger than the number of bytes I send to get this error.

**Comment by colbec** [comment]
I think I have seen something similar to the excessive bytes thing. I vaguely recall being asked if I had submitted big endian instead of little endian or something. Upshot was that I had sent the wrong package, wrong file, wrong sample rate possibly.

**Comment by Picolabo** [comment]
Do you remember if you needed to send a header before the frames, or just the frames? (this is what I am doing now: just the frames: 16kHz 16 bits PCM, by packets of something like 700 bytes)

**Comment by colbec** [comment]
Never operated at such a low level. Script functions have always taken care of that for me.

**Comment by Picolabo** [comment]
I have even tried to connect to Julius socket with telnet...
After the first 4 bytes I get the same error:

Error: rdwt: transfer data length exceeded: 8977504 (>640000)
STAT: no input frame
Stat: adin_tcpip: connection end

Hello M. Lee, I am lost into the C code, would you mind giving me a hint about the protocol between adintool and julius?

Best

**Comment by Picolabo** [comment]
I made it sending packets to adintool in server mode

**Comment by kunnetcn** [comment]
> I made it sending packets to adintool in server mode

Hi friends,
Did you found the answer about the question.I'm face the same trouble now...

**Comment by cajuncoder** [comment]
I am having this problem also. Please respond if you know anything about this.

**Comment by patrykplay** [comment]
Hi, someone solved this?

**Comment by MayamaTakeshi** [comment]
I got into this problem.
But i verified the adintool that comes with julius works without problems.
So I did some TCP packet captures to compare what adintool sends with the data that my app sends and I saw this:

adintool:
00000000  d0 07 00 00                                        ....
00000004  ff ff ff ff fe ff fe ff  fe ff 01 00 00 00 00 00   ........ ........
00000014  ff ff 00 00 01 00 00 00  ff ff 00 00 00 00 00 00   ........ ........
00000024  ff ff 00 00 ff ff fd ff  fd ff fe ff ff ff fd ff   ........ ........
00000034  fd ff ff ff 00 00 00 00  00 00 00 00 00 00 ff ff   ........ ........
00000044  fd ff ff ff fe ff ff ff  ff ff ff ff 00 00 01 00   ........ ........
00000054  03 00 02 00 01 00 01 00  01 00 00 00 ff ff 00 00   ........ ........


my_app:
00000000  ff ff ff ff fe ff fe ff  fe ff 01 00 00 00 00 00   ........ ........
00000010  ff ff 00 00 01 00 00 00  ff ff 00 00 00 00 00 00   ........ ........
00000020  ff ff 00 00 ff ff fd ff  fd ff fe ff ff ff fd ff   ........ ........
00000030  fd ff ff ff 00 00 00 00  00 00 00 00 00 00 ff ff   ........ ........
00000040  fd ff ff ff fe ff ff ff  ff ff ff ff 00 00 01 00   ........ ........
00000050  03 00 02 00 01 00 01 00  01 00 00 00 ff ff 00 00   ........ ........

notice that adintool sends 4 bytes before the actual data..
This is the length of the data segment (0x000007d0 = 2000) that will follow.
So just write this length to the socket before writing the actual data.
